### PR TITLE
Add missing field type DNI in brand address form

### LINKED ISSUE #16121: Unable to create new Brand Address when VAT identification number is mandatory
**Describe the bug**
If we have the VAT identification number required in countries, we cannot create a new Brand Address.

**To Reproduce**
1. Go to International -> Locations -> Countries
2. Edit Spain (España) or your country
3. Switch "Do you need a VAT identification number?" to yes.
4. Go to Catalog -> Brands & Suppliers -> Brands page
5. Create a new Brand Address with a country Spain
6. Save
7. see error

It is OK with PS1.7.5.2

**Screenshots**

image

**Additional information**
PrestaShop version: 1.7.6.1 / 1.7.6.x 
PHP version: 7.2

### PR DESCRIPTION
**Pull request by jolelievre**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | 1.7.6.x
| Description?  | Some countries require a DNI field, but it was missing in the Brand address form, so we add it in the form along with validation when it's required
| Type?         | bug fix
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Fixes #16121
| How to test?  | Try adding a brand address on a country which requires DNI (Spain for example) you can't add/update it if the field is empty (see issue for more details)

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/16153)

### CODE REVIEW

File: src/Core/ConstraintValidator/TypedRegexValidator.php [hunk 1]
```diff
@@ -87,6 +87,7 @@ private function getPattern($type)
             'message' => '/[<>{}]/i',
             'language_iso_code' => IsoCode::PATTERN,
             'language_code' => '/^[a-zA-Z]{2}(-[a-zA-Z]{2})?$/',
+            'dni_lite' => '/^[0-9A-Za-z-.]{1,16}$/U',
```
Comment (Progi1984) [inline src/Core/ConstraintValidator/TypedRegexValidator.php:90]: @jolelievre No way for creating a constant ?

File: src/Core/Domain/Address/Command/AddManufacturerAddressCommand.php [hunk 2]
```diff
@@ -121,7 +127,8 @@ public function __construct(
         $stateId = null,
         $homePhone = null,
         $mobilePhone = null,
-        $other = null
+        $other = null,
+        $dni = null
```
Comment (Progi1984) [inline src/Core/Domain/Address/Command/AddManufacturerAddressCommand.php:131]:
```suggestion
        ?string $dni = null
```

File: src/Core/Domain/Address/Command/AddManufacturerAddressCommand.php [hunk 3]
```diff
@@ -234,6 +242,14 @@ public function getOther()
         return $this->other;
     }
 
+    /**
+     * @return string|null
+     */
+    public function getDni()
```
Comment (Progi1984) [inline src/Core/Domain/Address/Command/AddManufacturerAddressCommand.php:248]:
```suggestion
    public function getDni(): ?string
```

File: src/Core/Domain/Address/Command/EditManufacturerAddressCommand.php [hunk 4]
```diff
@@ -320,6 +325,22 @@ public function setOther($other)
         $this->other = $other;
     }
 
+    /**
+     * @return string|null
+     */
+    public function getDni()
```
Comment (Progi1984) [inline src/Core/Domain/Address/Command/EditManufacturerAddressCommand.php:331]:
```suggestion
    public function getDni(): ?string
```

File: src/Core/Domain/Address/Command/EditManufacturerAddressCommand.php [hunk 5]
```diff
@@ -320,6 +325,22 @@ public function setOther($other)
...
+    {
+        return $this->dni;
+    }
+
+    /**
+     * @param string|null $dni
+     */
+    public function setDni($dni)
```
Comment (Progi1984) [inline src/Core/Domain/Address/Command/EditManufacturerAddressCommand.php:339]:
```suggestion
    public function setDni(?string $dni)
```

File: src/Core/Domain/Address/QueryResult/EditableManufacturerAddress.php [hunk 6]
```diff
@@ -127,7 +133,8 @@ public function __construct(
         $stateId,
         $homePhone,
         $mobilePhone,
-        $other
+        $other,
+        $dni
```
Comment (Progi1984) [inline src/Core/Domain/Address/QueryResult/EditableManufacturerAddress.php:137]:
```suggestion
        string $dni
```

File: src/Core/Domain/Address/QueryResult/EditableManufacturerAddress.php [hunk 7]
```diff
@@ -247,4 +255,12 @@ public function getOther()
     {
         return $this->other;
     }
+
+    /**
+     * @return string
+     */
+    public function getDni()
```
Comment (Progi1984) [inline src/Core/Domain/Address/QueryResult/EditableManufacturerAddress.php:262]:
```suggestion
    public function getDni(): string
```

File: src/PrestaShopBundle/Form/Admin/Sell/Address/ManufacturerAddressType.php [hunk 8]
```diff
@@ -95,8 +96,10 @@ public function buildForm(FormBuilderInterface $builder, array $options)
     {
         if ($options['country_id'] !== 0) {
             $countryIdForStateChoices = $options['country_id'];
+            $dniRequired = Country::isNeedDniByCountryId($options['country_id']);
         } else {
             $countryIdForStateChoices = $this->contextCountryId;
+            $dniRequired = Country::isNeedDniByCountryId($this->contextCountryId);
         }
```
Comment (Progi1984) [inline src/PrestaShopBundle/Form/Admin/Sell/Address/ManufacturerAddressType.php]:
```suggestion
        $countryIdForStateChoices = $options['country_id'] !== 0 ? $options['country_id'] : $this->contextCountryId;
        $dniRequired = Country::isNeedDniByCountryId($countryIdForStateChoices);
```

File: src/Core/Domain/Address/Command/AddManufacturerAddressCommand.php (hunk 2 above)
Comment (PierreRambaud) [reply src/Core/Domain/Address/Command/AddManufacturerAddressCommand.php:131]: It's 1.7.6.x, still compatible with 5.6. We can't type parameters.

File: src/Core/Domain/Address/Command/AddManufacturerAddressCommand.php (hunk 3 above)
Comment (PierreRambaud) [reply src/Core/Domain/Address/Command/AddManufacturerAddressCommand.php:248]: same we can't type parameters for 1.7.6.x

File: src/Core/Domain/Address/Command/AddManufacturerAddressCommand.php (hunk 2 above)
Comment (jolelievre) [reply src/Core/Domain/Address/Command/AddManufacturerAddressCommand.php:131]: We can't in 176 (PHP5)

File: src/Core/Domain/Address/Command/AddManufacturerAddressCommand.php (hunk 3 above)
Comment (jolelievre) [reply src/Core/Domain/Address/Command/AddManufacturerAddressCommand.php:248]: We can't in 176 (PHP5)

File: src/Core/Domain/Address/Command/EditManufacturerAddressCommand.php (hunk 4 above)
Comment (jolelievre) [reply src/Core/Domain/Address/Command/EditManufacturerAddressCommand.php:331]: We can't in 176 (PHP5)

File: src/Core/Domain/Address/Command/EditManufacturerAddressCommand.php (hunk 5 above)
Comment (jolelievre) [reply src/Core/Domain/Address/Command/EditManufacturerAddressCommand.php:339]: We can't in 176 (PHP5)

File: src/Core/Domain/Address/QueryResult/EditableManufacturerAddress.php (hunk 6 above)
Comment (jolelievre) [reply src/Core/Domain/Address/QueryResult/EditableManufacturerAddress.php:137]: We can't in 176 (PHP5)

File: src/Core/Domain/Address/QueryResult/EditableManufacturerAddress.php (hunk 7 above)
Comment (jolelievre) [reply src/Core/Domain/Address/QueryResult/EditableManufacturerAddress.php:262]: We can't in 176 (PHP5)

File: src/Core/ConstraintValidator/TypedRegexValidator.php (hunk 1 above)
Comment (jolelievre) [reply src/Core/ConstraintValidator/TypedRegexValidator.php:90]: As the other don't have const I preferred to do the same here (no related ValueObject to put the const) But maybe in the develop branch we can add it as I believe it's better organized

Comment (Progi1984) [reply src/Core/ConstraintValidator/TypedRegexValidator.php:90]: Accepted.

File: src/Adapter/Country/CountryDataProvider.php [hunk 9]
```diff
@@ -49,6 +51,24 @@ public function getCountries($id_lang, $active = false, $contain_states = false,
...
+     */
+    public function getCountriesIdWhichNeedDni()
+    {
+        $query = new DbQuery();
+        $query
+            ->select('c.`id_country`')
+            ->from('country', 'c')
+            ->where('c.`need_identification_number` = 1')
+        ;
+        $result =  Db::getInstance(_PS_USE_SQL_SLAVE_)->executeS($query);
```
Comment (Progi1984) [inline src/Adapter/Country/CountryDataProvider.php]: Double space after equal

File: src/Core/Form/ChoiceProvider/CountryByIdChoiceProvider.php [hunk 10]
```diff
@@ -44,6 +45,16 @@ final class CountryByIdChoiceProvider implements FormChoiceProviderInterface
      */
     private $langId;
 
+    /**
+     * @var array
+     */
+    private $countries;
+
+    /**
+     * @var array
```
Comment (Progi1984) [inline src/Core/Form/ChoiceProvider/CountryByIdChoiceProvider.php]:
```suggestion
     * @var array|int[]
```

File: src/PrestaShopBundle/Form/Admin/Sell/Address/ManufacturerAddressType.php [hunk 11]
```diff
@@ -94,10 +103,11 @@ public function __construct(
     public function buildForm(FormBuilderInterface $builder, array $options)
     {
         if ($options['country_id'] !== 0) {
-            $countryIdForStateChoices = $options['country_id'];
+            $countryId = $options['country_id'];
         } else {
-            $countryIdForStateChoices = $this->contextCountryId;
+            $countryId = $this->contextCountryId;
         }
```
Comment (Progi1984) [inline src/PrestaShopBundle/Form/Admin/Sell/Address/ManufacturerAddressType.php]:
```suggestion
        $countryId = $options['country_id'] !== 0 ? $options['country_id'] : $this->contextCountryId;
```

File: src/Adapter/Country/CountryDataProvider.php (hunk 9 above)
Comment (jolelievre) [reply src/Adapter/Country/CountryDataProvider.php]: Done

File: src/Core/Form/ChoiceProvider/CountryByIdChoiceProvider.php (hunk 10 above)
Comment (jolelievre) [reply src/Core/Form/ChoiceProvider/CountryByIdChoiceProvider.php]: Done

File: src/PrestaShopBundle/Form/Admin/Sell/Address/ManufacturerAddressType.php (hunk 11 above)
Comment (jolelievre) [reply src/PrestaShopBundle/Form/Admin/Sell/Address/ManufacturerAddressType.php]: Done

**Review by jolelievre — COMMENTED** [review]
Feedback addressed

**Review by Progi1984 — DISMISSED** [review]

**Comment by Robin-Fischer-PS** [comment]
Hi @jolelievre !

We've got a bug !

It's impossible to empty DNI field when it was filled before.

See video :
https://drive.google.com/file/d/1AHRBt8pQtXdDfdxUhpof7TF2e3LmUUZT/view?usp=sharing

Thanks :)

**Comment by jolelievre** [comment]
Hi @Robin-Fischer-PS 

Thank you, nice catch! Actually there was the same problem for other non-required fields (phone numbers, state, ...)

So I fixed it as well The internal problem is that the symfony form transforms empty string into `null` which was then ignored by the handler I detail this bug here because I'm sure similar bugs are present on other forms

**Review by Progi1984 — APPROVED** [review]

**Comment by Robin-Fischer-PS** [comment]
Thanks for the fix @jolelievre , it's OK now ! :smile:

**Pull request merged**